server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -333,6 +336,73 @@ public ShardAllocationDecision decideShardAllocation(ShardRouting shard, Routing
         assertThat(listenersCalled.get(), equalTo(1));
     }
 
+    public void testConcurrency() throws Exception {
+
+        var createdIndices = new CopyOnWriteArraySet<String>();
+        var allocatedIndices = new CopyOnWriteArraySet<String>();
+        var reroutedIndices = new CopyOnWriteArraySet<String>();
+
+        var threadPool = new TestThreadPool(getTestName());
+        RerouteService rerouteService = (r, p, l) -> threadPool.executor(ThreadPool.Names.CLUSTER_COORDINATION).submit(() -> {
+            reroutedIndices.addAll(createdIndices);
+            l.onResponse(null);
+        });
+        var allocator = new ShardsAllocator() {
+            @Override
+            public void allocate(RoutingAllocation allocation) {// TODO attempt to count up-to-date allocations
+                final var dataNodeId = allocation.nodes().getDataNodes().values().iterator().next().getId();
+                final var unassignedIterator = allocation.routingNodes().unassigned().iterator();
+                while (unassignedIterator.hasNext()) {
+                    String indexName = unassignedIterator.next().getIndexName();
+                    allocatedIndices.add(indexName);
+                    unassignedIterator.initialize(dataNodeId, null, 0L, allocation.changes());
+                }
```

Comment:
Can we simulate the case where the allocator can't allocate everything in one go here? I think this will allow for more interesting coverage of cases where the input goes stale mid-allocation.

```suggestion
                var madeProgress = false;
                while (unassignedIterator.hasNext()) {
                    final var indexName = unassignedIterator.next().getIndexName();
                    if (randomBoolean() || (madeProgress == false && unassignedIterator.hasNext() == false)) {
                        allocatedIndices.add(indexName);
                        unassignedIterator.initialize(dataNodeId, null, 0L, allocation.changes());
                        madeProgress = true;
                    } else {
                        unassignedIterator.removeAndIgnore(UnassignedInfo.AllocationStatus.NO_ATTEMPT, allocation.changes());
                    }
                }
```